contrib/raftexample/raft.go

```diff
@@ -428,6 +415,7 @@ func (rc *raftNode) serveChannels() {
 
 		// store raft entries to wal, then publish over commit channel
 		case rd := <-rc.node.Ready():
+			rc.maybeTriggerSnapshot()
```

Comment:
Why triggering snapshot it better here instead of just before .Advance ?

Reply:
We have to trigger snapshot before `publishEntries` because writing in-memory map is async via `commitC`. If not do so, when taking snapshot, entries may not be applied to the map yet.

Reply:
commitC is unbuffered channel. If 'nil' entry signaling end of 'Ready' is consumed it IMHO means that kvstore.readCommit applied all the changes into the in-memory map. 
I must be missing something.  

As this code is for educational purposes we need comments inline that makes it obvious to the readers.

Reply:
The `nil` entry on committees means the signal to load snapshot. So we cannot use it as an end of committed entries.
> As this code is for educational purposes we need comments inline that makes it obvious to the readers.

OK, I will add a comment.

Reply:
@ptabor My English is not so good. Please correct it if you feel it strange.

Reply:
> The `nil` entry on committees means the signal to load snapshot. So we cannot use it as an end of committed entries.

https://github.com/etcd-io/etcd/blob/ae7862e8bc8007eb396099db4e0e04ac026c8df5/contrib/raftexample/raft.go#L182

sends 'nil' always when 'Ready' processing is over. 

For majority of nils this branch triggers:

https://github.com/etcd-io/etcd/blob/ae7862e8bc8007eb396099db4e0e04ac026c8df5/contrib/raftexample/kvstore.go#L70

and they are just getting ignored.

Reply:
Yes, that was one of weird points in the previous code. No need to try to load snapshot every time. 
This comment if for my changes.
> The nil entry on committees means the signal to load snapshot. So we cannot use it as an end of committed entries.

Reply:
Forgive me digging in, but I still don't understand whether this move of line solves any problem. 

Intuitively generation of snapshot AFTER applying entries is mere comprehensive than generating snapshot BEFORE applying entries.

Reply:
> Intuitively generation of snapshot AFTER applying entries is mere comprehensive than generating snapshot BEFORE applying entries.

My idea is triggering snapshot first including applied entries of the previous Ready not this Ready. This snapshot doesn't include the committed entries of this Ready. 
But I realized that there is no guarantee that committed entries in previous Ready have already been applied in next Ready. This move line reduces the possibility of inconsistency, but doesn't solve the problem perfectly.

Reply:
I disagree for reasons from https://github.com/etcd-io/etcd/pull/11889#discussion_r568594007. 

Please show a repro of the 'inconsistency' problem or move the line back.

Reply:
IMHO, https://github.com/etcd-io/etcd/pull/11889#discussion_r568594007 is not related to this discussion...
To reproduce inconsistency, plz do following in this changes.
1.  Change `defaultSnapshotCount` to 5 and move maybeTriggerSnapshot to previous position.
2. run raftexample (goreman start) 
3. PUT 2 keys
    ```sh
    $ curl -L http://127.0.0.1:12380/my-key1 -XPUT -d bar
    $ curl -L http://127.0.0.1:12380/my-key2 -XPUT -d bar
    ```
4. restart raftexample
5. GET 2nd key
    ```sh
    $ curl -L http://127.0.0.1:12380/my-key2
    ```
Then you will get `Failed to GET`

Reply:
I found another possible solution like this from your comment. (Need to send nil twice for perfect guarantee?)
> commitC is unbuffered channel. 
```diff
diff --git a/contrib/raftexample/raft.go b/contrib/raftexample/raft.go
index 54f2cf7e9..31f32f99e 100644
--- a/contrib/raftexample/raft.go
+++ b/contrib/raftexample/raft.go
@@ -140,6 +140,7 @@ func (rc *raftNode) entriesToApply(ents []raftpb.Entry) (nents []raftpb.Entry) {
 // publishEntries writes committed log entries to commit channel and returns
 // whether all entries could be published.
 func (rc *raftNode) publishEntries(ents []raftpb.Entry) bool {
+	commitApplied := false
 	for i := range ents {
 		switch ents[i].Type {
 		case raftpb.EntryNormal:
@@ -153,6 +154,7 @@ func (rc *raftNode) publishEntries(ents []raftpb.Entry) bool {
 			case <-rc.stopc:
 				return false
 			}
+			commitApplied = true
 
 		case raftpb.EntryConfChange:
 			var cc raftpb.ConfChange
@@ -175,6 +177,14 @@ func (rc *raftNode) publishEntries(ents []raftpb.Entry) bool {
 		// after commit, update appliedIndex
 		rc.appliedIndex = ents[i].Index
 	}
+
+	if commitApplied {
+		select {
+		case rc.commitC <- nil:
+		case <-rc.stopc:
+			return false
+		}
+	}
 	return true
 }
 
@@ -417,7 +427,6 @@ func (rc *raftNode) serveChannels() {
 		case rd := <-rc.node.Ready():
 			// appliedIndex might be incremented before handling entries via commitC
 			// need to trigger snapshot for previous Ready here
-			rc.maybeTriggerSnapshot()
 
 			rc.wal.Save(rd.HardState, rd.Entries)
 			if !raft.IsEmptySnap(rd.Snapshot) {
@@ -431,6 +440,7 @@ func (rc *raftNode) serveChannels() {
 				rc.stop()
 				return
 			}
+			rc.maybeTriggerSnapshot()
 			rc.node.Advance()
 
 		case err := <-rc.transport.ErrorC:
```

I think this is more preferred because of this. HDYT?
> But I realized that there is no guarantee that committed entries in previous Ready have already been applied in next Ready. This move line reduces the possibility of inconsistency, but doesn't solve the problem perfectly.

Reply:
@ptabor Addressed. PTAL.